pool/opaque_test.go

```diff
@@ -0,0 +1,108 @@
+package pool
+
+import (
+	"reflect"
+	"runtime"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+type I interface {
+	Set(int)
+	Get() int
+}
+
+type T struct {
+	x I
+	p [4096]uint8
+}
+
+type N struct {
+	I
+	y int
+}
+
+func (n *N) Set(x int) {
+	n.y = x
+}
+
+func (n *N) Get() int {
+	return n.y
+}
+
+func TestOpaquePool(t *testing.T) {
+	i := NewOpaquePool(OpaquePoolOptions{
+		Size: 50000,
+		Type: reflect.TypeOf((*N)(nil)).Elem()})
+
+	p := NewOpaquePool(OpaquePoolOptions{
+		Construct: func(ptr interface{}) { ptr.(*T).x = i.Get().(I) },
```

Comment:
Yes. As we discussed, this works for pointers and interfaces, but unfortunately not for maps, slices or channels – these types allocate memory internally and we don't have control over it. For example, a slice would allocate its storage via `mallocgc()`, which will properly flag it on the heap, but GC will see it as unreachable and reclaim it since its ownership cannot be traced through our opaque heap.

I might make it work for "immutable" slices, e.g. by explicitly allocating slice storage from the opaque heap, populating the `reflect.SliceHeader` structure and casting it to a slice.

Maps and channels are complicated structures and are not exposed via `reflect`, so no luck here.